packages/server/src/loop.ts

```diff
@@ -34,160 +39,161 @@ export async function chargeCustomerInvoice({
+          populate: ['project', 'changes', 'customer'],
         },
       );
 
-      for await (let invoice of invoices) {
-        // Lock invoice processing
-        invoice.status = 'pending';
-        await database.em.persistAndFlush(invoice);
+      for await (const subscription of subscriptions) {
+        // TODO: should we lock subscription processing?
 
-        const { project, subscription } = invoice;
-        if (!subscription) {
-          log.error({ invoiceId: invoice._id }, 'Invoice has no subscription');
-          throw new Error('Invoice has no subscription');
-        }
+        const { project, customer } = subscription;
 
-        const { customer } = subscription;
+        const billingPeriod = getPeriodFromAnchorDate(subscription.nextPayment, subscription.anchorDate);
 
         try {
-          await database.em.populate(customer, ['activePaymentMethod']);
-          if (!customer.activePaymentMethod) {
-            log.error({ invoiceId: invoice._id, customerId: customer._id }, 'Customer has no active payment method');
-            throw new Error('Customer has no active payment method');
+          const existingInvoices = await database.invoices.find({
+            date: { $gte: billingPeriod.start, $lte: billingPeriod.end },
+            subscription,
+          });
+          if (existingInvoices.length > 0) {
+            log.error(
+              { subscriptionId: subscription._id, customerId: customer._id },
+              'Invoice for period already exists',
+            );
+            // TODO: should we just ignore this? currently this sets the subscription to error state?
+            throw new Error('Invoice for period already exists');
           }
 
-          invoice = addSubscriptionChangesToInvoice(subscription, invoice);
-          await chargeCustomerInvoice({ customer, invoice });
+          customer.invoiceCounter += 1;
+
+          const invoice = new Invoice({
+            currency: project.currency,
+            vatRate: project.vatRate,
+            sequentialId: customer.invoiceCounter,
+            subscription,
+            project,
+            status: 'draft',
+            date: billingPeriod.end, // TODO: or has this to be the current date when the invoice is issued?
+          });
+
+          const period = new SubscriptionPeriod(subscription, billingPeriod.start, billingPeriod.end);
+          period.getInvoiceItems().forEach((item) => {
+            invoice.items.add(item);
+          });
+
+          await database.em.persistAndFlush([customer, invoice]);
+
+          await chargeCustomerInvoice({ billingPeriod, customer, invoice });
+
+          const nextPeriod = getPeriodFromAnchorDate(
+            dayjs(subscription.nextPayment).add(1, 'day').toDate(),
+            subscription.anchorDate,
+          );
+          subscription.nextPayment = nextPeriod.end;
+          await database.em.persistAndFlush([subscription]);
```

Comment:
I currently decided to only update the `nextPayment` if the creation of the invoice and payment request was successful. Therefore if the creation fails the subscription would go into `error` state in will not be processed in upcoming months again until its manually fixed and the subscription will be considered inactive. If we would update the `nextPayment` anyways a user could have an inactive subscription for one month, but if the next month the error somehow would not happen again the subscription would be "recovered" however the payment for the failed month would simply be skipped and the user would not have the subscription for that month.